subprojects/normalization-java/src/main/java/org/gradle/api/internal/changedetection/state/LineEndingAwareResourceHasher.java

```diff
@@ -0,0 +1,199 @@
+        }
+    }
+
+    private static class BinaryContentDetectedException extends Exception {
+    }
+
+    @VisibleForTesting
+    static class LineEndingAwareInputStreamHasher {
+        private static final HashCode SIGNATURE = Hashing.signature(LineEndingAwareInputStreamHasher.class);
+        int peekAhead = -1;
+
+        public HashCode hash(InputStream inputStream) throws IOException, BinaryContentDetectedException {
+            byte[] buffer = new byte[8192];
+            PrimitiveHasher hasher = Hashing.newPrimitiveHasher();
+
+            hasher.putHash(SIGNATURE);
+            while (true) {
+                int nread = read(inputStream, buffer);
+                if (nread < 0) {
+                    break;
+                }
+
+                if (checkForControlCharacters(buffer, nread)) {
+                    throw new BinaryContentDetectedException();
+                }
+
+                hasher.putBytes(buffer, 0, nread);
+            }
+            return hasher.hash();
+        }
+
+        int read(InputStream inputStream, byte[] b) throws IOException {
+            byte[] original = new byte[b.length];
+
+            // If there is something left over in the peekAhead buffer, use that as the first byte
+            int readLimit = b.length;
+            if (peekAhead != -1) {
+                readLimit--;
+            }
+
+            int index = 0;
+            int read = inputStream.read(original, 0, readLimit);
+            if (read != -1) {
+                Iterator<Byte> itr = Bytes.asList(original).subList(0, read).iterator();
+                while (itr.hasNext()) {
+                    // Get our next byte from the peek ahead buffer if it contains anything
+                    int next = peekAhead;
+
+                    // If there was something in the peek ahead buffer, use it, otherwise get the next byte
+                    if (next != -1) {
+                        peekAhead = -1;
+                    } else {
+                        next = itr.next();
+                    }
+
+                    // If the next bytes are '\r' or '\r\n', replace it with '\n'
+                    if (next == '\r') {
+                        peekAhead = itr.hasNext() ? itr.next() : inputStream.read();
+                        if (peekAhead == '\n') {
+                            peekAhead = -1;
+                        }
+                        next = '\n';
+                    }
+
+                    b[index++] = (byte) next;
+                }
+            } else if (peekAhead != -1) {
+                // If there is a character still left in the peekAhead buffer but not in the input stream, then normalize it and return
+                int next = peekAhead;
+                peekAhead = -1;
+                if (next == '\r') {
+                    next = '\n';
+                }
+                b[index++] = (byte) next;
+            }
+            return index == 0 ? -1 : index;
+        }
```

Comment:
This feels a bit like the reimplementation of `BufferedInputStream`'s `read()`. I wonder if we could simplify this.

Reply:
I've refactored this a bit to use `ByteBuffer` and simplified the logic a bit.  It's a little less complex now.